fontconfig: fix etc priority

ericsagnes opened the pull request:
###### Motivation for this change

Address #16026

---

Current state issues:
- ~~When /etc/fonts is already present, the new `/etc/fonts` will be created as `/etc/fonts.tmp` breaking fontconfig configuration. Issue #16978~~ fixed by #17042
- ~~using `symlinkJoin` for creating the `/etc/font` package make `test.installers.*` fails because `lndir` is not available. Issue #16983~~ fixed by using `buildEnv` instead of `symlinkJoin`

---

This make all the fontconfig related configuration files be symlinked in the correct order in `etc` so fontconfig settings and `fontconfig.defaultFonts` are applied correctly.

This will trigger a rebuild of fontconfig and all related packages, so the PR is made on `staging`.

The change in fontconfig is required in order to have a correct file loading order for `${pkgs.fontconfig.out}/etc/fonts/conf.d/*`. (that are currently linked in `/etc/fonts/{,2.11}/fonts.conf`).

example:

```
$ FC_DEBUG=1024 fc-match serif
FC_DEBUG=1024
    Loading config file /etc/fonts/2.11/fonts.conf
    Scanning config dir /etc/fonts/2.11/conf.d
    Loading config file /etc/fonts/2.11/conf.d/00-nixos.conf
    Loading config file /etc/fonts/2.11/conf.d/10-nixos-rendering.conf
    Loading config file /etc/fonts/2.11/conf.d/10-scale-bitmap-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/20-unhint-small-vera.conf
    Loading config file /etc/fonts/2.11/conf.d/30-liberation-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/30-liberation-sans-narrow.conf
    Loading config file /etc/fonts/2.11/conf.d/30-liberation-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/30-liberation-serif.conf
    Loading config file /etc/fonts/2.11/conf.d/30-metric-aliases-free.conf
    Loading config file /etc/fonts/2.11/conf.d/30-metric-aliases.conf
    Loading config file /etc/fonts/2.11/conf.d/30-nixos-generic-alias.conf
    Loading config file /etc/fonts/2.11/conf.d/30-urw-aliases.conf
    Loading config file /etc/fonts/2.11/conf.d/31-titillium.conf
    Loading config file /etc/fonts/2.11/conf.d/36-repl-missing-glyphs.conf
    Loading config file /etc/fonts/2.11/conf.d/37-repl-global-free.conf
    Loading config file /etc/fonts/2.11/conf.d/37-repl-webfonts.conf
    Loading config file /etc/fonts/2.11/conf.d/40-alef.conf
    Loading config file /etc/fonts/2.11/conf.d/40-amiri.conf
    Loading config file /etc/fonts/2.11/conf.d/40-arphic-ukai.conf
    Loading config file /etc/fonts/2.11/conf.d/40-arphic-uming.conf
    Loading config file /etc/fonts/2.11/conf.d/40-ddc-uchen.conf
    Loading config file /etc/fonts/2.11/conf.d/40-dejavusans-yuanti-condensed.conf
    Loading config file /etc/fonts/2.11/conf.d/40-dejavusans-yuanti-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/40-dejavusans-yuanti.conf
    Loading config file /etc/fonts/2.11/conf.d/40-dzongkha.conf
    Loading config file /etc/fonts/2.11/conf.d/40-faruma.conf
    Loading config file /etc/fonts/2.11/conf.d/40-hannom.conf
    Loading config file /etc/fonts/2.11/conf.d/40-himalaya-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/40-ipafont.conf
    Loading config file /etc/fonts/2.11/conf.d/40-ipamjfont.conf
    Loading config file /etc/fonts/2.11/conf.d/40-khmer-os.conf
    Loading config file /etc/fonts/2.11/conf.d/40-koruri.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lklug.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-assamese.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-bengali.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-devanagari.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-gujarati.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-gurmukhi.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-kannada.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-malayalam.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-marathi.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-nepali.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-odia.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-punjabi.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-tamil.conf
    Loading config file /etc/fonts/2.11/conf.d/40-lohit-telugu.conf
    Loading config file /etc/fonts/2.11/conf.d/40-melthofonts.conf
    Loading config file /etc/fonts/2.11/conf.d/40-mph-2b-damase.conf
    Loading config file /etc/fonts/2.11/conf.d/40-mplus.conf
    Loading config file /etc/fonts/2.11/conf.d/40-myanmar3.conf
    Loading config file /etc/fonts/2.11/conf.d/40-nanum-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/40-nanumgothic-coding.conf
    Loading config file /etc/fonts/2.11/conf.d/40-non-latin-microsoft.conf
    Loading config file /etc/fonts/2.11/conf.d/40-non-latin-misc.conf
    Loading config file /etc/fonts/2.11/conf.d/40-nonlatin.conf
    Loading config file /etc/fonts/2.11/conf.d/40-noto-arabic.conf
    Loading config file /etc/fonts/2.11/conf.d/40-sawarabi.conf
    Loading config file /etc/fonts/2.11/conf.d/40-saweri.conf
    Loading config file /etc/fonts/2.11/conf.d/40-source-han-sans-cn.conf
    Loading config file /etc/fonts/2.11/conf.d/40-source-han-sans-jp.conf
    Loading config file /etc/fonts/2.11/conf.d/40-source-han-sans-kr.conf
    Loading config file /etc/fonts/2.11/conf.d/40-source-han-sans-twhk.conf
    Loading config file /etc/fonts/2.11/conf.d/40-tharlon.conf
    Loading config file /etc/fonts/2.11/conf.d/40-umeplus.conf
    Loading config file /etc/fonts/2.11/conf.d/40-unfonts-core.conf
    Loading config file /etc/fonts/2.11/conf.d/40-vlgothic.conf
    Loading config file /etc/fonts/2.11/conf.d/40-wqy-microhei.conf
    Loading config file /etc/fonts/2.11/conf.d/40-wqy-zenhei.conf
    Loading config file /etc/fonts/2.11/conf.d/43-wqy-zenhei-sharp.conf
    Loading config file /etc/fonts/2.11/conf.d/45-aboriginal-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-aboriginal-serif.conf
    Loading config file /etc/fonts/2.11/conf.d/45-aileron.conf
    Loading config file /etc/fonts/2.11/conf.d/45-alegreya-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-alegreya.conf
    Loading config file /etc/fonts/2.11/conf.d/45-aleo.conf
    Loading config file /etc/fonts/2.11/conf.d/45-amble.conf
    Loading config file /etc/fonts/2.11/conf.d/45-andada.conf
    Loading config file /etc/fonts/2.11/conf.d/45-antonio.conf
    Loading config file /etc/fonts/2.11/conf.d/45-archivo-black.conf
    Loading config file /etc/fonts/2.11/conf.d/45-archivo-narrow.conf
    Loading config file /etc/fonts/2.11/conf.d/45-arev-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-arsenal.conf
    Loading config file /etc/fonts/2.11/conf.d/45-berenis-adf-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/45-bitter.conf
    Loading config file /etc/fonts/2.11/conf.d/45-bravo.conf
    Loading config file /etc/fonts/2.11/conf.d/45-buenard.conf
    Loading config file /etc/fonts/2.11/conf.d/45-caladea.conf
    Loading config file /etc/fonts/2.11/conf.d/45-camingocode.conf
    Loading config file /etc/fonts/2.11/conf.d/45-cantarell.conf
    Loading config file /etc/fonts/2.11/conf.d/45-cantoraone.conf
    Loading config file /etc/fonts/2.11/conf.d/45-carlito.conf
    Loading config file /etc/fonts/2.11/conf.d/45-caviar-dreams.conf
    Loading config file /etc/fonts/2.11/conf.d/45-charis-sil.conf
    Loading config file /etc/fonts/2.11/conf.d/45-clear-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-comme.conf
    Loading config file /etc/fonts/2.11/conf.d/45-consola-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/45-cooper-hewitt.conf
    Loading config file /etc/fonts/2.11/conf.d/45-courier-prime.conf
    Loading config file /etc/fonts/2.11/conf.d/45-crimson-text.conf
    Loading config file /etc/fonts/2.11/conf.d/45-croscore-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/45-dejavu.conf
    Loading config file /etc/fonts/2.11/conf.d/45-droid.conf
    Loading config file /etc/fonts/2.11/conf.d/45-ebgaramond.conf
    Loading config file /etc/fonts/2.11/conf.d/45-enriqueta.conf
    Loading config file /etc/fonts/2.11/conf.d/45-erewhon.conf
    Loading config file /etc/fonts/2.11/conf.d/45-exo2.conf
    Loading config file /etc/fonts/2.11/conf.d/45-fanwood.conf
    Loading config file /etc/fonts/2.11/conf.d/45-fira-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/45-fira-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-gelasio.conf
    Loading config file /etc/fonts/2.11/conf.d/45-gentium.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica-ce-35-thin.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica-ce-55-roman.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica-cy.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica-narrow.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica-neue-lt-com.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica-neue.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica-world.conf
    Loading config file /etc/fonts/2.11/conf.d/45-helvetica.conf
    Loading config file /etc/fonts/2.11/conf.d/45-heuristica.conf
    Loading config file /etc/fonts/2.11/conf.d/45-inconsolata-pwl.conf
    Loading config file /etc/fonts/2.11/conf.d/45-inconsolatazi4.conf
    Loading config file /etc/fonts/2.11/conf.d/45-infini.conf
    Loading config file /etc/fonts/2.11/conf.d/45-inknut-antiqua.conf
    Loading config file /etc/fonts/2.11/conf.d/45-iosevka.conf
    Loading config file /etc/fonts/2.11/conf.d/45-istok.conf
    Loading config file /etc/fonts/2.11/conf.d/45-junicode.conf
    Loading config file /etc/fonts/2.11/conf.d/45-latin-microsoft.conf
    Loading config file /etc/fonts/2.11/conf.d/45-latin-misc.conf
    Loading config file /etc/fonts/2.11/conf.d/45-latin-modern.conf
    Loading config file /etc/fonts/2.11/conf.d/45-latin.conf
    Loading config file /etc/fonts/2.11/conf.d/45-lato.conf
    Loading config file /etc/fonts/2.11/conf.d/45-league-gothic.conf
    Loading config file /etc/fonts/2.11/conf.d/45-lekton.conf
    Loading config file /etc/fonts/2.11/conf.d/45-liberastika.conf
    Loading config file /etc/fonts/2.11/conf.d/45-liberation.conf
    Loading config file /etc/fonts/2.11/conf.d/45-libre-baskerville.conf
    Loading config file /etc/fonts/2.11/conf.d/45-libre-caslon.conf
    Loading config file /etc/fonts/2.11/conf.d/45-linden-hill.conf
    Loading config file /etc/fonts/2.11/conf.d/45-linux-libertine-o.conf
    Loading config file /etc/fonts/2.11/conf.d/45-linux-libertine.conf
    Loading config file /etc/fonts/2.11/conf.d/45-lobster-two.conf
    Loading config file /etc/fonts/2.11/conf.d/45-lora.conf
    Loading config file /etc/fonts/2.11/conf.d/45-luxi.conf
    Loading config file /etc/fonts/2.11/conf.d/45-magra.conf
    Loading config file /etc/fonts/2.11/conf.d/45-merriweather-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-merriweather.conf
    Loading config file /etc/fonts/2.11/conf.d/45-monoid.conf
    Loading config file /etc/fonts/2.11/conf.d/45-montserrat.conf
    Loading config file /etc/fonts/2.11/conf.d/45-noticia-text.conf
    Loading config file /etc/fonts/2.11/conf.d/45-noto-cros.conf
    Loading config file /etc/fonts/2.11/conf.d/45-noto-nonlatin.conf
    Loading config file /etc/fonts/2.11/conf.d/45-noto-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-noto-serif.conf
    Loading config file /etc/fonts/2.11/conf.d/45-opensans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-oswald.conf
    Loading config file /etc/fonts/2.11/conf.d/45-overpass.conf
    Loading config file /etc/fonts/2.11/conf.d/45-oxygen.conf
    Loading config file /etc/fonts/2.11/conf.d/45-paratype.conf
    Loading config file /etc/fonts/2.11/conf.d/45-permian.conf
    Loading config file /etc/fonts/2.11/conf.d/45-pfeffer-mediaeval.conf
    Loading config file /etc/fonts/2.11/conf.d/45-pfennig.conf
    Loading config file /etc/fonts/2.11/conf.d/45-playfair-display.conf
    Loading config file /etc/fonts/2.11/conf.d/45-quintessential.conf
    Loading config file /etc/fonts/2.11/conf.d/45-raleway.conf
    Loading config file /etc/fonts/2.11/conf.d/45-roboto-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/45-roboto.conf
    Loading config file /etc/fonts/2.11/conf.d/45-sansation.conf
    Loading config file /etc/fonts/2.11/conf.d/45-scada.conf
    Loading config file /etc/fonts/2.11/conf.d/45-sen.conf
    Loading config file /etc/fonts/2.11/conf.d/45-signika.conf
    Loading config file /etc/fonts/2.11/conf.d/45-sinkin-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-sorts-mill-goudy.conf
    Loading config file /etc/fonts/2.11/conf.d/45-source-code-pro-pwl.conf
    Loading config file /etc/fonts/2.11/conf.d/45-source-code-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/45-source-sans-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/45-source-serif-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/45-symbola.conf
    Loading config file /etc/fonts/2.11/conf.d/45-tehuti.conf
    Loading config file /etc/fonts/2.11/conf.d/45-tex-gyre.conf
    Loading config file /etc/fonts/2.11/conf.d/45-titillium.conf
    Loading config file /etc/fonts/2.11/conf.d/45-triod-postnaja.conf
    Loading config file /etc/fonts/2.11/conf.d/45-ubuntu.conf
    Loading config file /etc/fonts/2.11/conf.d/45-unb-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/45-urw-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/45-vera-humana-95.conf
    Loading config file /etc/fonts/2.11/conf.d/45-vollkorn.conf
    Loading config file /etc/fonts/2.11/conf.d/45-weblysleek-ui.conf
    Loading config file /etc/fonts/2.11/conf.d/45-work-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/45-yanone-kaffeesatz.conf
    Loading config file /etc/fonts/2.11/conf.d/49-sansserif.conf
    Loading config file /etc/fonts/2.11/conf.d/52-fontconfig-ultimate.conf
    Loading config file /etc/fonts/2.11/conf.d/60-latin-free.conf
    Loading config file /etc/fonts/2.11/conf.d/60-latin.conf
    Loading config file /etc/fonts/2.11/conf.d/65-fonts-persian.conf
    Loading config file /etc/fonts/2.11/conf.d/65-non-latin-free.conf
    Loading config file /etc/fonts/2.11/conf.d/65-nonlatin.conf
    Loading config file /etc/fonts/2.11/conf.d/65-noto-sans-ui.conf
    Loading config file /etc/fonts/2.11/conf.d/65-noto-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/65-noto-serif.conf
    Loading config file /etc/fonts/2.11/conf.d/65-ttf-droid-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/65-ttf-droid-serif.conf
    Loading config file /etc/fonts/2.11/conf.d/65-wqy-zenhei.conf
    Loading config file /etc/fonts/2.11/conf.d/66-aliases-wine-free.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-assamese.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-bengali.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-devanagari.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-gujarati.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-gurmukhi.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-kannada.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-malayalam.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-marathi.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-nepali.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-odia.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-punjabi.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-tamil.conf
    Loading config file /etc/fonts/2.11/conf.d/66-lohit-telugu.conf
    Loading config file /etc/fonts/2.11/conf.d/67-override-aliases.conf
    Loading config file /etc/fonts/2.11/conf.d/68-override.conf
    Loading config file /etc/fonts/2.11/conf.d/69-unifont.conf
    Loading config file /etc/fonts/2.11/conf.d/80-delicious.conf
    Loading config file /etc/fonts/2.11/conf.d/88-forced-synthetic.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-aileron.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-alegreya-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-alegreya.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-aleo.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-andada.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-arsenal.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-berenis-adf-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-bitter.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-bravo.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-buenard.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-cantarell.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-cantoraone.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-cooper-hewitt.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-crimson-text.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-dejavu.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-ebgaramond.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-enriqueta.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-erewhon.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-exo2.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-fanwood.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-fira-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-fira-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-helvetica-neue-lt-com.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-helvetica-neue.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-helvetica.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-heuristica.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-inconsolata-pwl.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-inconsolatazi4.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-infini.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-latin-modern.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-league-gothic.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-libre-baskerville.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-libre-caslon.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-linden-hill.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-linux-libertine.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-lobster-two.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-luxi.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-magra.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-melthofonts.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-montserrat.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-mplus.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-oswald.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-paratype.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-permian.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-pfeffer-mediaeval.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-quintessential.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-raleway.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-sawarabi.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-scada.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-sen.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-signika.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-sinkin-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-sorts-mill-goudy.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-code-pro-pwl.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-code-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-han-sans-cn.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-han-sans-jp.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-han-sans-kr.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-han-sans-twhk.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-sans-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-source-serif-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-tehuti.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-tex-gyre.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-titillium.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-unb-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-urw-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-vollkorn.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-work-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-wqy-microhei.conf
    Loading config file /etc/fonts/2.11/conf.d/90-non-tt-yanone-kaffeesatz.conf
    Loading config file /etc/fonts/2.11/conf.d/90-synthetic.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-aboriginal-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-aboriginal-serif.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-alef.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-alegreya-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-alegreya.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-amble.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-amiri.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-andada.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-antonio.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-archivo-black.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-archivo-narrow.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-arev-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-arphic-ukai.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-arphic-uming.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-caladea.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-camingocode.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-cantoraone.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-carlito.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-caviar-dreams.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-charis-sil.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-clear-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-comme.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-consola-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-courier-prime.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-crimson-text.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-croscore-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-ddc-uchen.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-dejavu.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-dejavusans-yuanti-condensed.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-dejavusans-yuanti-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-dejavusans-yuanti.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-droid.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-dzongkha.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-ebgaramond.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-faruma.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-fonts-microsoft.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-fonts-misc.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-gelasio.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-gentium.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-hannom.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica-ce-35-thin.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica-ce-55-roman.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica-cy.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica-narrow.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica-neue-lt-com.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica-neue.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica-world.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-helvetica.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-heuristica.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-himalaya-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-inknut-antiqua.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-iosevka.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-ipafont.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-ipamjfont.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-istok.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-junicode.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-khmer-os.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-koruri.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lato.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lekton.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-liberastika.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-liberation.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-libre-baskerville.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-libre-caslon.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-linux-libertine.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lklug.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-assamese.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-bengali.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-devanagari.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-gujarati.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-gurmukhi.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-kannada.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-malayalam.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-marathi.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-nepali.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-odia.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-punjabi.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-tamil.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lohit-telugu.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-lora.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-luxi.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-merriweather-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-merriweather.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-meslo.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-monoid.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-mph-2b-damase.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-myanmar3.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-nanum-fonts.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-nanumgothic-coding.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-noticia-text.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-noto-cros.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-noto-nonlatin.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-noto-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-noto-serif.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-opensans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-overpass.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-oxygen.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-paratype.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-pfennig.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-playfair-display.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-raleway.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-roboto-mono.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-roboto.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-sansation.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-saweri.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-signika.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-sinkin-sans.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-sorts-mill-goudy.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-source-code-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-source-sans-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-source-serif-pro.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-symbola.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-tharlon.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-triod-postnaja.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-ubuntu.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-umeplus.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-unfonts-core.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-vera-humana-95.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-vlgothic.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-vollkorn.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-weblysleek-ui.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-wqy-zenhei.conf
    Loading config file /etc/fonts/2.11/conf.d/90-tt-yanone-kaffeesatz.conf
    Loading config file /etc/fonts/2.11/conf.d/92-selective-rendering-microsoft.conf
    Loading config file /etc/fonts/2.11/conf.d/92-selective-rendering-misc.conf
    Loading config file /etc/fonts/2.11/conf.d/93-final-rendering.conf
    Loading config file /etc/fonts/2.11/conf.d/94-no-synthetic.conf
    Loading config file /etc/fonts/2.11/conf.d/95-reject.conf
    Loading config file /etc/fonts/2.11/conf.d/97-selective-rendering-custom.conf
    Loading config file /etc/fonts/2.11/conf.d/99-user.conf
SourceSerifPro-Regular.otf: "Source Serif Pro" "Regular"
```
###### Notes
- When `/etc/fonts` is already present, the new folder will be created as `/etc/fonts.tmp` unless `/etc/fonts` is manually deleted. I suppose it is because current  `/etc/fonts` is a real folder and this PR `/etc/fonts` is a symlink.
- I am not sure about the `/etc/fonts` and `/etc/fonts/2.11` difference, I tried to keep it as close as it was before the PR, but I would appreciate if someone with better fontconfig knowledge could verify everything work as intended.
- it seems that when multiple are set, the order `fontconfig.defaultFonts` is affected by the system locale (verified with asian locales and japanese fonts), as far as I can tell it seems to be a feature of fontconfig to raise locale matching fonts at a higher priority than what they are declared.
- fontconfig documentation being quite vague and still full of mysteries to me, there should room for improvement .

Tests, feedback and improvements are welcome.

cc @vcunat @ttuegel 
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bjornfor commented:
For whom will loosing fontconfig < 2.11 be a problem? Non-NixOS users on old systems? (If so, how old?)

abbradar commented:
I think Ubuntu 12.04 LTS has old fontconfig and is still supported upstream. Also SteamOS derives from it, and so Steam uses this old version (2.8.0) too.

vcunat commented:
Old systems shouldn't be broken, as my old comments #2050 claim that newer fontconfig library doesn't have a problem with old `/etc/fonts`. IIRC our changes only fix the combination of old nix-built binaries on new NixOS. *But the whole 2.11 discussion seems not that much relevant to this PR.*

abbradar commented:
Indeed and indeed. BTW, Hydra has not yet started evaluating several staging merges. They all are quite focused in what they really impact, so I propose batching this patch, too. I'll test it in a VM and if it's okay -- merge.

vcunat commented:
I'm currently testing four stdenv-rebuilding changes on top of staging, so I certainly count on those in this iteration. I haven't yet got to properly reviewing this PR.

abbradar commented:
I have reviewed it some time ago, but it's difficult to catch all the changes because there's lots of moving lines around in play. I think it looks okay (there is an unreferenced derivation for testing left, but it can be fixup'ed). Testing in VM showed that all `fontconfig` settings now do apply, Infinality patch works and `conf.d` is properly populated. I think this is good to go, but I don't want to rush things too much -- if you think it needs more review or some of your staging patches overlap in scope with this, let's wait!

abbradar commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -2,6 +2,124 @@
+      "ln -s ${userConf}        $out/etc/fonts/${version}/conf.d/99-user.conf"}
+      
+    '';
+    confPkg2 = pkgs.runCommand "test" {} ''
```
This is leftover from testing IIUC.

ericsagnes replied:
Nice catch! Thanks for noticing, I updated the PR.

ericsagnes commented:
Closing as this has been merged in https://github.com/NixOS/nixpkgs/commit/1e53d4a7776acbf61f42c094c103652c8068ad64.

ericsagnes commented:
Reopening as this was reverted by cfc0a54.
prepending title with WIP.

Current state issues:
- When `/etc/fonts` is already present, the new `/etc/fonts` will be created as `/etc/fonts.tmp` breaking fontconfig configuration. Issue #16978
- using `symlinkJoin` for creating the `/etc/font` package make `test.installers.*` fails because `lndir` is not available. Issue #16983

ericsagnes commented:
OK, the PR is becoming a real mess.

So this was based on staging, and it was merged but for some reason the PR is still open.
Then staging was merged to master and then reverted by cfc0a54 on master only.

So it is applied on staging, applied and reverted on master, and I wrongly rebased it on master so there are hundred something commits that don't belong here.

But as the same time as it is not reverted on staging it generates conflicts if I try to apply it on the latest staging.

What can be done from here? revert 1e53d4a on this PR and adding the new changes on top of that?

abbradar commented:
@ericsagnes I've merged master to staging -- if you rebase you can then revert the revert and make your fixes.

FWIW I've tried to fix #16983  yesterday but no luck -- when I add `lndir` to the closure I get weird evaluation errors instead.

ericsagnes committed:
fontconfig: fix etc priority

ericsagnes commented:
@abbradar Thank you!
I cleaned up the commit and switched `symlinkJoin` for `buildEnv` so there should be no need for `lndir`.

But `nix-build '<nixos/release.nix>' -A tests.installer.simple` fails with some random `anonymous function at /tmp/foo called without required argument 'bar'` that seems unrelated.

abbradar commented:
@ericsagnes That's what I means by "weird evaluation errors". Interesting thing is that I managed to get several different seemingly unrelated errors. But I don't remember how ;_;

rasendubi commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -142,136 +338,18 @@ with lib;
+    fonts.fontconfig.confPkgs = [ confPkg ];
+
+    # Proposition A: use environment.etc
+    # TO FIX: this this create /etc/fonts.tmp if /etc/fonts is already present
```
Created a fix at https://github.com/NixOS/nixpkgs/pull/17042.

Please, take a look.

ericsagnes replied:
Great! thanks, I will try that.

ericsagnes replied:
Works perfectly, thank you!

rasendubi commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -142,136 +338,18 @@ with lib;
+    environment.etc.fonts.source = "${fontconfigEtc}/etc/fonts/";
+
+    # Proposition B: use environment.pathsToLink
+    # TO FIX: For some reason this does not to not link /etc/fonts/ correctly
```
Don't know what "not correctly" here, but that might be fixed at https://github.com/NixOS/nixpkgs/pull/16828.

abbradar commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -339,17 +344,10 @@ in
   };
   config = mkIf cfg.enable {
-    fonts.fontconfig.confPkgs = [ confPkg ];
+    fonts.fontconfig.confPkgs    = [ confPkg ];
```
A nitpick: we usually name options like those `*Packages`, not `*Pkgs` -- for example, `udev.packages`, `environment.systemPackages`, `boot.kernelPackages` etc.

ericsagnes replied:
Good catch, I updated the option name.
Thanks!

ericsagnes commented:
As #17042 was merged, this PR is ready to be merged.
Since the previous merge, lot of cleanup happened so it should be easier to review. 

The core functionality of the module is not changed, just a few improvements were made:
- put all fontconfig configuration file under etc so configuration files are parsed in the right order (see PR description for detail)
- cache file, previous `00-nixos.conf`, is now also generated for the default fontconfig (`support` version).
- `generic-alias.conf` priority was changed from 60 to 30 so default fonts should always be at the head of the preferred font list.
- cleanup and comments

The installer problem should be fixed as `symlinkJoin` was changed to `buildEnv`, but `tests.installer.simple` keeps failing with what seems unrelated reasons on my machine (even when using a vanilla 16.03).

The PR was tested it with multiple font setups on a VM and a real machine.

Balletie commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -1,7 +1,203 @@
+
+    # user settings configuration file
+    # priority 99
+    userConf = pkgs.writeText "fc-99-user.conf" ''
```
In the default way fontconfig is set up, this has priority 50 (see https://cgit.freedesktop.org/fontconfig/tree/conf.d/50-user.conf).

But `50-user.conf` is removed in line 69 of fontconfig/default.nix.

Balletie commented on pkgs/development/libraries/fontconfig/default.nix:
```diff
@@ -68,7 +68,6 @@ stdenv.mkDerivation rec {
     cd "$out/etc/fonts"
     rm conf.d/{50-user,51-local}.conf
```
I know you didn't change this part, but why is this removed? Why are all font configurations in `configuration.nix` or wherever not converted to XML in `/etc/fonts/local.conf` as is the default?

I think a lot the problems we are having with priorities are due to this. And now that our fontconfig setup is different from other major OSes, the documentation of fontconfig configuration on e.g. the arch wiki or elsewhere on the internet does not apply to NixOS.

ericsagnes replied:
Thank for reviewing! 
Indeed, that is a good question, the commit comment for this change was "add an upstream bugfix, remove other config inclusion". (27968ece)
Maybe @vcunat can bring more insight about this choice?

vcunat replied:
I don't think that commit is important. I just didn't want to invest enough time for a truly proper solution, like this PR attempts (I suppose). (At that time it's possible I didn't even know that fontconfig doesn't interleave the directories.)

ericsagnes replied:
I see, thanks for the answer.

After looking more at it, `51-local.conf` is removed because it is the file we want to generate via the module (`30-nixos-generic-alias.conf` in this PR).

I guess `50-user.conf` is removed and generated in the module for interaction with `includeUserConf` setting, but that could be improved.

I will look if a commit can be added to make this PR closer to upstream definitions.

ericsagnes replied:
@Balletie I moved `30-nixos-generic-alias.conf` to the `local.conf` 51 priority, it work almost perfect beside that `fontconfig-ultimate`'s [`43-wqy-zenhei-sharp.conf`](https://github.com/bohoomil/fontconfig-ultimate/blob/master/fontconfig_patches/fonts-settings/43-wqy-zenhei-sharp.conf) set deja-vu and WenQuanYi as preferred fonts, taking priority over rules defined in `local.conf`.

I submitted an upstream [issue](https://github.com/bohoomil/fontconfig-ultimate/issues/173), and propose that in the meantime we patch fontconfig-ultimate to move `43-wqy-zenhei-sharp.conf` to `60-wqy-zenhei-sharp.conf` as it is a generic->family configuration.

Do you think this is an acceptable solution?

Balletie commented on pkgs/development/libraries/fontconfig/make-fonts-conf.xsl:
```diff
@@ -28,8 +28,6 @@
       <cachedir>/var/cache/fontconfig</cachedir>
 
-      <!-- fontconfig distribution conf.d -->
-      <include><xsl:value-of select="$fontconfig" />/etc/fonts/conf.d</include>
```
For reference, the above line is now done by lines 167 through 171 in `fontconfig.nix`: https://github.com/ericsagnes/nixpkgs/blob/da64fed0a378e243b3c50847308a29fd9667810e/nixos/modules/config/fonts/fontconfig.nix#L167-L171

ericsagnes committed:
fontconfig module: respect upstream definitions

ericsagnes committed:
fontconfig-ultimate: fix wqy-zenhei-sharp priority

ericsagnes commented:
Added 2 commits.

### fontconfig module: respect upstream definitions

This keep upstream `50-user.conf` and `51-local.conf` files.
`51-local.conf` load the `/etc/font/local.conf` file.

`/etc/font/local.conf` content is automatically generated with the contents of the new `localConf` module option.

`localConf` can be used for any global fontconfig setting that cannot be set via other module options.

The module default fonts configuration priority as been moved to 52, with a lower priority than upstream user (50) and local (51) configuration files.


### fontconfig-ultimate: fix wqy-zenhei-sharp priority

A file from fontconfig-ultimate font patch set, wqy-zenhei-sharp, has wrong priority (43) and set default fonts with a higher priority than any "alternate config file loading" files making `defaultFont` settings meaningless. 

This minor patch fix the file priority so `defaultFonts` settings work as expected.

Balletie commented:
Yes, this already looks much better to me :).

Balletie commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -179,15 +170,21 @@ let cfg = config.fonts.fontconfig;
+      # local.conf (indirect priority 51)
+      ${optionalString (cfg.localConf != "") ''
+      ln -s ${localConf}        $support_folder/local.conf
+      ln -s ${localConf}        $latest_folder/local.conf
```
I wonder whether this gets included anywhere, since `latest_folder` is the versioned directory (e.g. `/etc/fonts/2.11`)?

ericsagnes replied:
Nice catch!
Indeed the latest `local.conf` wasn't used.

Even if both `local.conf` have the same contents, I amended the commit so latest `51-local.conf` to look for latest `local.conf`.  (https://github.com/NixOS/nixpkgs/pull/16730/commits/da342b0b017a44b36c536870ba4b75b18a2c2ebf#diff-376672cba12c87d4103e6134632e306fR164)

It feels cleaner and it will make things easier if in the future we want to have a different `local.conf` for each version.

Balletie replied:
> It feels cleaner and it will make things easier if in the future we want to have a different `local.conf` for each version.

Weren't the versioned configurations only put in place for backwards compatibility?

ericsagnes replied:
> Weren't the versioned configurations only put in place for backwards compatibility?

Right, I must confess that I don't know much about the reasons and implications for multiple fontconfig version support so I might have made silly assumptions.

I just thought that there might be a need to have different `local.conf` for each fontconfig version in case there were incompatibilities.

But the main reason is that, as a personal taste, I think it easier to understand and cleaner to have symmetric file structure for both versions.

But I have no problem in removing `latest/local.conf` in favor of a single `/etc/fonts/local.conf` if it is more appropriate.

Balletie replied:
Nah it's fine, that makes sense.

abbradar commented:
I made [a branch](https://github.com/abbradar/nixpkgs/tree/font-updates) for combined testing of new font-related updates (#16730, #17770, #17846). I believe this is ready for merging, so after #17846 is also ready I'll test them all, keep for a few more days in case anyone has objections and merge to staging. That said, I may have skipped some unresolved problem arised in this thread -- are there any?

ericsagnes commented:
@abbradar 

> That said, I may have skipped some unresolved problem arised in this thread -- are there any?

There shouldn't be anymore problems.

The only thing that I couldn't check is the `tests.installer.simple` test working fine as it keeps failing on my test machine for random reasons (even when using stable channels).

In theory, the `symlinkJoin` -> `buildEnv` change should fix the problem raised in https://github.com/NixOS/nixpkgs/issues/16983.

ericsagnes commented:
Thanks for merging!

abbradar commented:
You are welcome, thanks for your contributions ^_^
-- 
Nikolay. 

Sent from a mobile device, please excuse my brevity.
